### PR TITLE
Add support for multiple versions of RHEL

### PR DESCRIPTION
**Pull request by cottsay**
This change adds a RHEL version parameter to the CI job infrastructure and updates the RHEL Dockerfile to support RHEL 9.

The changes to the Dockerfile include three packages being downloaded directly from the Fedora build servers while we wait for them to make their way through testing. I see no reason to merge this change prior to that, so I'm opening it as a draft until those packages go stable and we can drop the conditionals.

I have no idea how adding the new Jenkins parameter will impact reruns of existing jobs or the launcher job or anything - I'm not even really sure how to test it. Seeking guidance.

I made all of the conditionals "optimistic", meaning that RHEL versions other than 8 or 9 should take the desired path (i.e. omit workarounds and include more system packages).

RHEL 8 Test CI: [Build Status](https://ci.ros2.org/view/All/job/test_ci_linux-rhel/13/) (Same 3 test failures as the nightly build)
RHEL 9 Test CI: [Build Status](https://ci.ros2.org/view/All/job/test_ci_linux-rhel/11/) (One CMake warning, known issue with RHEL 9)

### CODE REVIEW

File: linux_docker_resources/Dockerfile-RHEL [hunk 1]
```diff
@@ -59,24 +60,37 @@ RUN dnf update \
...
+  fi
 
 # These dependencies were generated by invoking rosdep on the latest version of the target workspace:
 # rosdep install --from-paths src --ignore-src --simulate --reinstall --skip-keys ...
 RUN dnf install \
     CUnit-devel \
     acl \
-    assimp-devel \
+    $(if test ${EL_RELEASE} != 8; then echo asio-devel; fi) \
+    $(if test ${EL_RELEASE} != 9; then echo assimp-devel; fi) \
```
Comment (clalancette) [inline linux_docker_resources/Dockerfile-RHEL:82]: I know you were trying to make this opportunistic, but it feels like this scatters around a lot of if statements.  Might it make more sense to three blocks of `dnf install` commands:

1. Package names that are common to both EL8 and EL9
2. Package names that are only for EL8
3. Package names that are only for EL9

?

File: linux_docker_resources/Dockerfile-RHEL [hunk 2]
```diff
@@ -1,9 +1,10 @@
-FROM almalinux:8
+FROM almalinux:9
 ARG ROS_DISTRO=rolling
+ARG EL_RELEASE=9
```
Comment (clalancette) [inline linux_docker_resources/Dockerfile-RHEL:3]: I don't think this matters too much, but: don't we want to leave the default as 8 for now?  That is, we aren't quite ready to switch over to 9 yet.

**Review by clalancette — CHANGES_REQUESTED** [review]
I've left a couple of comments inline.  I also have no idea what this is going to do with existing jobs.

For testing it, couldn't we do a deploy to the test buildfarm?

File: linux_docker_resources/Dockerfile-RHEL (hunk 2 above)
Comment (cottsay) [reply linux_docker_resources/Dockerfile-RHEL]: We're pretty darn close, but you're right. Done in 4c042d2.

**Comment by cottsay** [comment]
> For testing it, couldn't we do a deploy to the test buildfarm?

It sounds like that isn't a viable option right now. I deployed to the `test_ci_linux-rhel` jobs on [ci.ros2.org](https://ci.ros2.org/) for now. I'll look into adding a `test_ci_launcher` job to validate that interaction. That should give us sufficient validation coverage.

File: linux_docker_resources/Dockerfile-RHEL (hunk 1 above)
Comment (cottsay) [reply linux_docker_resources/Dockerfile-RHEL:82]: Eh, I'm lukewarm about this. These inline conditionals aren't particularly long or verbose (especially as far as bash is concerned), and it helps to avoid creating new and short-lived blocks.

I find this approach more maintainable, and I'll probably be doing most of the maintenance. Not a hill I'm willing to die on, but I'd prefer it.

**Comment by cottsay** [comment]
As far as validation is concerned, I confirmed that when re-running a build of a job that was executed prior to the introduction of the new parameter, the new parameter value in the rerun takes the default value specified here (so RHEL 8).

Also, specifying the extra parameter in the launcher job doesn't break old child jobs that don't have the new parameter.

tl;dr - Jenkins seems to be doing the "right thing" wherever I check corner cases. I'm not sure what else to try, but I'm open to suggestions.

File: linux_docker_resources/Dockerfile-RHEL (hunk 1 above)
Comment (clalancette) [reply linux_docker_resources/Dockerfile-RHEL:82]: I don't feel that strongly about it, so if you prefer this, that is OK with me.

**Comment by clalancette** [comment]
> tl;dr - Jenkins seems to be doing the "right thing" wherever I check corner cases. I'm not sure what else to try, but I'm open to suggestions.

That sounds like enough validation to me.  https://ci.ros2.org gets looked at pretty often, so if we deploy and something goes wrong, we'll notice pretty quickly and can take action.

**Review by clalancette — COMMENTED** [review]
The code here looks OK to me.  My only remaining issue has to do with the download of bullet and pydot directly from Koji.  I know you are working to make those full packages, so I think we should probably wait to merge this until that happens.  Thoughts?

**Comment by cottsay** [comment]
> I think we should probably wait to merge this until that happens.

Agreed, we should wait. They're both on QA, so it shouldn't be long.

**Comment by cottsay** [comment]
> 

This is done - the packages are stable.

**Review by clalancette — APPROVED** [review]

**Comment by cottsay** [comment]
This has been deployed.
